### ISSUE TITLE
What about a scikit-surgerytensorflow

### ISSUE THREAD

**Issue by thompson318** [UCL/scikit-surgery#25]
In GitLab by @MattClarkson on Nov 21, 2018, 14:51

@StephenThompson @ThomasDowrick - this ticket to discuss how to integrate learned models into SNAPPY.

So, in my simple view of the world, you'd have:

* Researcher hacks around, implements network, trains network on data... at some point decides 'it works'
* So, to make widely available you would need
    * package that has tensor flow as dependency, and all associated libraries, so could be quite big.
    * something describing network
    * something describing weights
    * some harness, to say, apply it to an openCV image.

So, a question to SNAPPY developers - shall we get someone to lead this, as in practice we have enough to do. Also, the mechanics of training, so some harness to run on a load of data could be quite involved. It could also be quite specific to each dataset. So, Im not sure how scaleable it would be. But in the first instance, Bongjin has trained MaskR-CNN on Eli's old data set so we have an exemplar on something to start with.

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Nov 21, 2018, 15:39

With the disclaimer that I only have a beginner's knowledge of tensorflow, I have a few thoughts:

*  Will there be enough commonality between different ML projects to allow for code reuse?
*  Is there functionality that isn't already in tensorflow that would warrant creating a new module just for IGI?
*  If we think the workflow could be quite specific to each set of data, is it better to have the user implement this aspect of the code separately for each project, in the same way that we think GUIs will individual to each project?
*  If we do think a separate package is needed, could this also include scikit-learn (or other ML packages), or just tensorflow?

Logistically, if there is someone who is available and willing to start working on a new library, then I'd be fine for them to start trying. Between the three of us, there is probably enough to do in the meantime polishing/completing the existing list of packages before we start on new ones.

**Comment by thompson318** [comment]
In GitLab by @MattClarkson on Nov 21, 2018, 16:17

I'll ask Bongjin, and see if there is a need, if not, we can park it for now.

@BongjinKoo

**Comment by thompson318** [comment]
In GitLab by @BongjinKoo on Nov 22, 2018, 09:25

All issues raised above are valid.
And, I am worried that this might become just like NiftyNet,
which is a(n unnecessary) wrapper around TensorFlow (TF)
that not many people use (from my perspective).
More so, given that we usually find a (state-of-the-art) network implementation,
from the authors, TensorFlow, or somewhere else, for our purpose,
and extend/modify it.
So, I think we can just let the users work on integrating their TF code into SNAPPY
as Tom said. 
At least for now.

**Comment by thompson318** [comment]
In GitLab by @MattClarkson on Nov 22, 2018, 11:22

unassigned @StephenThompson

**Comment by thompson318** [comment]
In GitLab by @MattClarkson on Nov 22, 2018, 11:22

OK. Nice to have discussion. Leave it for now. I'll close, we can re-open if necessary.

**Comment by thompson318** [comment]
In GitLab by @MattClarkson on Nov 22, 2018, 11:22

closed
